File: android/src/main/kotlin/io/sentient/android/chat/message/ChatRows.kt
// ChatRows — fold messages into a render list with day dividers (webui .day-divider).
// A divider precedes the first message of each calendar day; label = day + first ts.
package io.sentient.android.chat.message

import io.sentient.mobiledata.outbox.PendingMessage
import io.sentient.mobilesdk.sdk.ChatMessage
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

sealed interface ChatRow {
    data class Divider(val label: String, val key: String) : ChatRow
    data class Msg(val message: ChatMessage, val index: Int) : ChatRow

    /**
     * An optimistic pending user message from the outbox — shown with a status chip
     * (QUEUED / SENT / FAILED) until the gateway echoes the committed entry back and
     * ChatRepository reconciles it away by pendingId.
     */
    data class Pending(val msg: PendingMessage) : ChatRow
}

/**
 * Stable per-message LazyColumn key — ONE row per logical BUBBLE across its
 * whole lifecycle. Keyed on replyId FIRST, not turnId: a mid-turn steer rotates
 * replyId, so one turnId can own TWO assistant bubbles (reply 1 answers the
 * first message, reply 2 answers the steer). Keying on turnId aliases both
 * replies onto the same LazyColumn key — worse than iOS here, Compose THROWS on
 * a duplicate key rather than just mis-animating.
 *
 * The live streaming bubble and its committed twin still share replyId
 * (ObserveChatUseCase builds the live bubble with replyId = it.replyId), so the
 * streaming→committed handoff is still the SAME row (grows in place, no
 * remount). replyId is constant across tokens same as turnId was, so unlike ts
 * it never churns mid-reveal.
 *
 * Entries with no replyId (user rows, REST history) key by their stable gateway
 * entryId. turnId is a fallback only for a gateway that does not stamp replyId.
 * Index is the last resort. Mirrors iOS ChatRow.
 */
internal fun messageRowKey(m: ChatMessage, index: Int): String = when {
    // The optimistic row and its live committed echo deliberately share identity.
    m.role == "user" && !m.pendingId.isNullOrEmpty() -> "send-${m.pendingId}"
    !m.replyId.isNullOrEmpty() -> "reply-${m.replyId}"
    m.entryId.isNotEmpty() -> "ent-${m.entryId}"
    !m.turnId.isNullOrEmpty() -> "turn-${m.turnId}"
    else -> "idx-$index"
}

fun chatRows(messages: List<ChatMessage>, nowMs: Long): List<ChatRow> {
    val out = ArrayList<ChatRow>(messages.size + 4)
    var lastKey: String? = null
    for ((i, m) in messages.withIndex()) {
        // Streaming messages have ts=0 (no real timestamp while in flight).
        // Never bucket them into a day-divider — ts=0/epoch would produce a
        // bogus "Thursday, Jan 1 1970" separator. The committed entry that
        // follows carries the real ts and its own divider. Mirrors iOS ChatRows.swift.
        if (!m.streaming) {
            val key = dayKey(m.ts)
            if (key != lastKey) { out.add(ChatRow.Divider(dividerLabel(m.ts, nowMs), key)); lastKey = key }
        }
        out.add(ChatRow.Msg(m, i))
    }
    return out
}

private fun dayKey(ts: Long): String {
    val c = Calendar.getInstance().apply { timeInMillis = ts }
    return "${c.get(Calendar.YEAR)}-${c.get(Calendar.DAY_OF_YEAR)}"
}

private fun dividerLabel(ts: Long, nowMs: Long): String {
    // Calendar-based yesterday (DST-aware) — a raw `nowMs - 86_400_000L` mislabels
    // the boundary on 23h/25h DST-transition days. Mirrors iOS isDateInYesterday.
    val yesterdayMs = Calendar.getInstance().apply { timeInMillis = nowMs; add(Calendar.DATE, -1) }.timeInMillis
    val day = when (dayKey(ts)) {
        dayKey(nowMs) -> "Today"
        dayKey(yesterdayMs) -> "Yesterday"
        else -> SimpleDateFormat("EEEE", Locale.getDefault()).format(Date(ts))
    }
    val time = SimpleDateFormat("h:mm a", Locale.getDefault()).format(Date(ts))
    return "$day · $time"
}

File: ios/App/Chat/message/ChatRows.swift
// ChatRows — the typed internal chronology for the chat message surface.
//
// The chronology is the single source for committed messages, optimistic
// outbox rows, and day dividers. It preserves the gateway identity contract
// while keeping pending rows after authoritative history and out of date
// grouping.
import Foundation
import MobileData

enum MessageChronologyRow: Identifiable {
    case divider(label: String, id: String)
    case message(ChatMessage, index: Int, continuation: Bool)
    case pending(PendingMessage, index: Int)

    var id: String {
        switch self {
        case let .divider(_, id): return "div-\(id)"
        case let .message(message, index, _): return messageRowId(message, index: index)
        case let .pending(message, _): return "send-\(message.id)"
        }
    }
}

/// Compatibility name for existing internal/test callers. New rendering code
/// consumes `MessageChronologyRow` through `messageChronology`.
typealias ChatRow = MessageChronologyRow

struct MessageChronology {
    let rows: [MessageChronologyRow]
    /// Number of visible message rows, excluding day dividers. This is the
    /// chronology total used by the shared accessibility label.
    let messageCount: Int
}

/// Stable per-message identity — ONE row per logical bubble across its whole
/// lifecycle. Reveal projection marks its retained presentation identity in
/// entryId so the turn placeholder, first stamped delta, and committed echo
/// remain one native row. Other assistant rows key on replyId: a mid-turn steer
/// rotates replyId, so separate actual replies remain separate rows.
///
/// Entries with no replyId (user rows, REST history) key by their stable
/// gateway entryId. turnId is a fallback only for a gateway that does not
/// stamp replyId. Index is the last resort.
func messageRowId(_ message: ChatMessage, index: Int) -> String {
    // Optimistic and committed user echoes share the pendingId identity.
    if message.role == "user", let pendingId = message.pendingId, !pendingId.isEmpty {
        return "send-\(pendingId)"
    }
    if message.entryId.hasPrefix("presentation:") { return message.entryId }
    if let replyId = message.replyId, !replyId.isEmpty { return "reply-\(replyId)" }
    if !message.entryId.isEmpty { return "ent-\(message.entryId)" }
    if let turnId = message.turnId, !turnId.isEmpty { return "turn-\(turnId)" }
    return "idx-\(index)"
}

/// Fold authoritative messages and transient outbox entries into the one
/// render list consumed by MessageList. Pending entries are appended after
/// committed history, receive no day divider, and never become continuations.
/// A pending entry whose committed echo is already present is omitted as a
/// boundary guard; the data layer normally filters it first, but this keeps a
/// transient echo race from rendering duplicate bubbles.
func messageChronology(
    messages: [ChatMessage],
    pending: [PendingMessage] = [],
    calendar: Calendar = .current,
    now: Date = Date()
) -> MessageChronology {
    var rows: [MessageChronologyRow] = []
    var lastDay: DateComponents?
    var previous: ChatMessage?

    for (index, message) in messages.enumerated() {
        var insertedDivider = false
        // Live rows carry their stable start timestamp for identity metadata,
        // but remain outside durable day grouping until committed.
        if !message.streaming {
            let date = Date(timeIntervalSince1970: Double(message.ts) / 1000)
            let day = calendar.dateComponents([.year, .month, .day], from: date)
            if day != lastDay {
                let dayKey = "\(day.year ?? 0)-\(day.month ?? 0)-\(day.day ?? 0)"
                rows.append(.divider(
                    label: dividerLabel(date, calendar: calendar, now: now),
                    id: dayKey
                ))
                lastDay = day
                insertedDivider = true
            }
        }

        let continuation = !insertedDivider && previous?.role == message.role
        rows.append(.message(message, index: index, continuation: continuation))
        previous = message
    }

    let echoedPendingIds = Set(messages.compactMap { message -> String? in
        guard message.role == "user", let pendingId = message.pendingId, !pendingId.isEmpty else {
            return nil
        }
        return pendingId
    })
    let visiblePending = pending.filter { !echoedPendingIds.contains($0.id) }
    for (offset, message) in visiblePending.enumerated() {
        rows.append(.pending(message, index: messages.count + offset))
    }

    return MessageChronology(rows: rows, messageCount: messages.count + visiblePending.count)
}

/// Compatibility wrapper for the pre-encapsulation name. It delegates to the
/// single chronology implementation and intentionally has no second policy.
func chatRows(_ messages: [ChatMessage], calendar: Calendar = .current,
              now: Date = Date()) -> [ChatRow] {
    messageChronology(messages: messages, calendar: calendar, now: now).rows
}

private let dividerDayFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "EEEE"
    return formatter
}()

private let dividerTimeFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.timeStyle = .short
    formatter.dateStyle = .none
    return formatter
}()

private func dividerLabel(_ date: Date, calendar: Calendar, now: Date) -> String {
    let day: String
    if calendar.isDateInToday(date) { day = "Today" }
    else if calendar.isDateInYesterday(date) { day = "Yesterday" }
    else { day = dividerDayFormatter.string(from: date) }
    return "\(day) · \(dividerTimeFormatter.string(from: date))"
}
